// @ethereumjs/block/dist/cjs/from-beacon-payload.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executionPayloadFromBeaconPayload = void 0;
const util_1 = require("@ethereumjs/util");
/**
 * Converts a beacon block execution payload JSON object {@link BeaconPayloadJson} to the {@link ExecutionPayload} data needed to construct a {@link Block}.
 * The JSON data can be retrieved from a consensus layer (CL) client on this Beacon API `/eth/v2/beacon/blocks/[block number]`
 */
function executionPayloadFromBeaconPayload(payload) {
    const executionPayload = {
        parentHash: payload.parent_hash,
        feeRecipient: payload.fee_recipient,
        stateRoot: payload.state_root,
        receiptsRoot: payload.receipts_root,
        logsBloom: payload.logs_bloom,
        prevRandao: payload.prev_randao,
        blockNumber: (0, util_1.bigIntToHex)(BigInt(payload.block_number)),
        gasLimit: (0, util_1.bigIntToHex)(BigInt(payload.gas_limit)),
        gasUsed: (0, util_1.bigIntToHex)(BigInt(payload.gas_used)),
        timestamp: (0, util_1.bigIntToHex)(BigInt(payload.timestamp)),
        extraData: payload.extra_data,
        baseFeePerGas: (0, util_1.bigIntToHex)(BigInt(payload.base_fee_per_gas)),
        blockHash: payload.block_hash,
        transactions: payload.transactions,
    };
    if (payload.withdrawals !== undefined && payload.withdrawals !== null) {
        executionPayload.withdrawals = payload.withdrawals.map((wd) => ({
            index: (0, util_1.bigIntToHex)(BigInt(wd.index)),
            validatorIndex: (0, util_1.bigIntToHex)(BigInt(wd.validator_index)),
            address: wd.address,
            amount: (0, util_1.bigIntToHex)(BigInt(wd.amount)),
        }));
    }
    if (payload.blob_gas_used !== undefined && payload.blob_gas_used !== null) {
        executionPayload.blobGasUsed = (0, util_1.bigIntToHex)(BigInt(payload.blob_gas_used));
    }
    if (payload.excess_blob_gas !== undefined && payload.excess_blob_gas !== null) {
        executionPayload.excessBlobGas = (0, util_1.bigIntToHex)(BigInt(payload.excess_blob_gas));
    }
    if (payload.parent_beacon_block_root !== undefined && payload.parent_beacon_block_root !== null) {
        executionPayload.parentBeaconBlockRoot = payload.parent_beacon_block_root;
    }
    if (payload.execution_witness !== undefined && payload.execution_witness !== null) {
        // the casing structure in payload is already camel case, might be updated in
        // kaustinen relaunch
        executionPayload.executionWitness = payload.execution_witness;
    }
    return executionPayload;
}
exports.executionPayloadFromBeaconPayload = executionPayloadFromBeaconPayload;
